### PR TITLE
Akoumparouli/low mem mixtral ckpt converter

### PR DESCRIPTION
**Pull request by akoumpa**
# What does this PR do ?
This add the option --low-ram to enable checkpoint conversion in low-ram computers for large checkpoints (e.g. 250GB).

Memory reduction is achieved by not having the full converted checkpoint and the huggingface checkpoint on memory at the same time. Instead, it saves the converted checkpoint layer by layer, remove the HF checkpoint from memory and finally restores the converted layers and saves to .nemo format.

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: scripts/checkpoint_converters/convert_mixtral_hf_to_nemo.py [hunk 1]
```diff
@@ -195,6 +207,15 @@
...
     trainer = Trainer(plugins=plugins, accelerator='cpu', precision=precision, strategy=NLPDDPStrategy())
+    return trainer, dtype
+
+
+def convert(args):
+    logging.info(f"loading checkpoint {args.input_name_or_path}")
+
+    model_args, ckpt, tokenizer = load_mixtral_ckpt(args.input_name_or_path)
+    nemo_config = load_config(model_args, tokenizer.vocab_file)
+    trainer, dtype = make_trainer(args, nemo_config)
```
[the remark this answers is not in the document]
Comment (akoumpa) [reply scripts/checkpoint_converters/convert_mixtral_hf_to_nemo.py]: removed

**Comment by akoumpa** [comment]
jenkins

**Comment by akoumpa** [comment]
jenkins

File: scripts/checkpoint_converters/convert_mixtral_hf_to_nemo.py [hunk 2]
```diff
@@ -108,6 +113,9 @@ def load_config(mixtral_config, tokenizer_path):
     # RMSNorm's epsilon.
     nemo_config.layernorm_epsilon = mixtral_config['rms_norm_eps']
     nemo_config.normalization = 'rmsnorm'
+    nemo_config.micro_batch_size = 1
```
Comment (yaoyu-33) [inline scripts/checkpoint_converters/convert_mixtral_hf_to_nemo.py:116]: why this is necessary?

Comment (akoumpa) [reply scripts/checkpoint_converters/convert_mixtral_hf_to_nemo.py:116]: was getting a micro_batch_size of 4 in the final config for some reason.

File: scripts/checkpoint_converters/convert_mixtral_hf_to_nemo.py [hunk 3]
```diff
@@ -342,5 +393,21 @@ def convert(args):
 
 if __name__ == '__main__':
     args = get_args()
+    if args.low_ram:
+        os.makedirs(args.tmp_dir, exist_ok=True)
+
     parallel_state.set_expert_model_parallel_world_size(1)
-    convert(args)
+    checkpoint = OrderedDict()
```
Comment (yaoyu-33) [inline scripts/checkpoint_converters/convert_mixtral_hf_to_nemo.py:400]: Can you add some comments how the below change helps low ram? 
- load partial and save partial => then load and merge? how did it save mem

Comment (akoumpa) [reply scripts/checkpoint_converters/convert_mixtral_hf_to_nemo.py:400]: Sure, memory reduction is achieved by not having the full converted checkpoint and the huggingface checkpoint on memory concurrently, instead you save the converted checkpoint layer by layer and then remove HF checkpoint from memory and then final save to .nemo format.

**Review by yaoyu-33 — APPROVED** [review]

**Pull request merged**